compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1005,7 +1033,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let mut coerce = CoerceMany::with_coercion_sites(coerce_to, args);
             assert_eq!(self.diverges.get(), Diverges::Maybe);
             for e in args {
-                let e_ty = self.check_expr_with_hint(e, coerce_to);
+                let e_ty = self.check_expr_with_expectation(
+                    e,
+                    ExpectHasType(coerce_to, expected.get_coercion_ctxt()),
```

Comment:
Hmm, so this is interesting... I think this is another case where we would really want to thread down the `CoerceMany` instance. What's going on here is that we have something like:

```rust
let x: [T] = [a, b, c];
```

I think that having the `check_expr_coercable_to` call occur inside `check_expr_with_expectation` and *then* calling `coerce.coerce` may cause a problem. Hmm.